js/ui/status/keyboard.js

```diff
@@ -480,6 +483,49 @@ const InputSourceManager = new Lang.Class({
             this._updateMruSettings();
     },
 
+    get passwordModeEnabled() {
+        return this._passwordModeEnabled;
+    },
+
+    set passwordModeEnabled(enable) {
+        if (this._passwordModeEnabled === enable)
+            return;
+
+        this._passwordModeEnabled = enable;
+        this._syncPasswordMode();
+    },
+
+    _syncPasswordMode: function() {
+        let source;
+
+        if (this._passwordModeEnabled) {
+            // do not switch the keyboard layout if the current one is a latin one
+            if (this._keyboardManager.isLatinLayout(this._currentSource.id))
+                return;
+
+            // look for latin keyboard layout to switch to
+            for (let i = 0; i < this._mruSources.length; ++i) {
+                let src = this._mruSources[i];
+                if (this._keyboardManager.isLatinLayout(src.id)) {
+                    source = src;
+                    break;
+                }
+            }
+            if (!source) {
+                // fallback to an 'en' keyboard layout
+                this._updateInputSources(true);
+                source = this._fallbackInputSource;
+            }
+            this._lastSource = this._currentSource;
```

Comment:
Is `this._lastSource` still needed right now? We don't seem to do anything with it anymore. Related question: does the "transient" latin source go away from the system configuration when the dialog is dismissed? (It should.)